keras_cv/training/contrastive/contrastive_trainer.py

```diff
@@ -0,0 +1,260 @@
+        classes: optional number of classes to classify images into, only to be
+            specified if `include_probe` is True.
+
+    Returns:
+      A `keras.Model` instance.
+
+
+    Usage:
+    ```python
+    encoder = keras_cv.models.DenseNet121(include_rescaling=False, include_top=False, pooling="avg")
+    augmenter = keras_cv.layers.preprocessing.RandomFlip()
+    projector = keras.layers.Dense(64)
+
+    trainer = keras_cv.training.ContrastiveTrainer(
+        encoder,
+        augmenter,
+        projector
+    )
+
+    trainer.compile(
+        optimizer=keras.optimizers.Adam(),
+        loss=keras_cv.losses.SimCLRLoss()
+    )
+
+    unlabeled_images = load_data()
+    trainer.fit(unlabeled_images)
+    ```
+
+    """
+
+    def __init__(
+        self,
+        encoder,
+        augmenter,
+        projector,
+        include_probe,
+        classes=None,
+    ):
+        super().__init__()
+
+        if encoder.output.shape.rank != 2:
+            raise ValueError("Encoder must have a flattened output")
+
+        if include_probe:
+            if not classes:
+                raise ValueError(
+                    "`classes` must be specified when `include_probe` is `True`."
+                )
+
+        if type(augmenter) is tuple and len(augmenter) != 2:
+            raise ValueError(
+                "`augmenter` must be either a single augmenter or a tuple of exactly 2 augmenters."
+            )
+
+        if type(projector) is tuple and len(projector) != 2:
+            raise ValueError(
+                "`augmenter` must be either a single augmenter or a tuple of exactly 2 augmenters."
+            )
+
+        self.augmenters = (
+            augmenter if type(augmenter) is tuple else (augmenter, augmenter)
+        )
+        self.encoder = encoder
+        self.projectors = (
+            projector if type(projector) is tuple else (projector, projector)
+        )
+
+        self.loss_metric = keras.metrics.Mean(name="loss")
+
+        self.include_probe = include_probe
+
+        if self.include_probe:
+            self.probing_top = layers.Dense(classes, name="linear_probe")
+            self.probe_loss_metric = keras.metrics.Mean(name="probe_loss")
+            self.probe_metrics = []
+
+    def compile(
+        self, probe_optimizer=None, probe_loss=None, probe_metrics=None, **kwargs
+    ):
+        super().compile(**kwargs)
+
+        if self.include_probe and not probe_optimizer:
+            raise ValueError(
+                "`probe_optimizer` must be specified when `include_probe` is `True`."
+            )
+
+        if self.include_probe and not probe_loss:
+            raise ValueError(
+                "`probe_loss` must be specified when `include_probe` is `True`."
+            )
+
+        if self.include_probe:
+            self.probe_loss = probe_loss
+            self.probe_optimizer = probe_optimizer
+            self.probe_metrics = probe_metrics or []
```

Comment:
you can create a static method to produce a linear probe to make creating probes easier:

approximately...
```
@staticmethod
def linear_probe(classes, **kwargs):
    return Sequential(Dense(classes)), **kwargs)
```